### PR TITLE
ci(jenkins): release process

### PR DESCRIPTION
**Pull request by v1v**
## Highlights
- enable ITs: async when running as a PR or sync when running in a branch/tag
- Continuous delivery when running the pipeline either in a branch or tag. master branch refers to `latest` while others will name the image with their names.
- Script to bump versions for the Node.js agent.

## Issues
- Depends on:
  - https://github.com/elastic/apm-integration-testing/pull/556
  - https://github.com/elastic/apm-pipeline-library/pull/271
  - https://github.com/elastic/apm-agent-nodejs/pull/1553

## Tasks
- [x] Versioning
- [x] Release automation
- [x] ITs validation should be synchronous when running a release.

### CODE REVIEW

File: .ci/bump-version.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  -e HOME=/tmp \
+  -e AGENT_VERSION="${AGENT_VERSION}" \
+  -w /app \
+  -v "$(pwd):/app" \
+  node:12-alpine /bin/sh -c "set -x
+    CI=true npm install elastic-apm-node@${AGENT_VERSION}"
+
+# Commit changes
+git add package.json
+git commit -m "Bump version ${AGENT_VERSION}"
```
Comment (watson) [inline .ci/bump-version.sh]: To not confuse with bumping the version of opbeans-node, but instead upgrading a dependency, I'd change this line to:

```suggestion
git commit -m "fix(package): bump elastic-apm-node to v${AGENT_VERSION}"
```

**Comment by v1v** [comment]
> I don't think you need to run npm install again as long as you just ensure to build a new Docker image as the version dependency is already specified as a range. If we do that you also don't risk upgrading to a version with a higher major component as that will be a breaking change.

The idea of this particular approach is to ensure the version for the apm-agent-nodejs which has been shipped got an opbeans docker image with it. Therefore we got reproducibility, using ranges might be good for certain cases, but for this particular one, there is a requirement to ensure all the releases for all the apm-agents got a specific opbeans docker image matching with the same version.

The next follow-ups will consist of running some functional tests to validate whether those dependencies are not breaking anything. The one which will do that is the:
- https://github.com/elastic/apm-integration-testing/blob/master/.ci/scripts/opbeans.sh

It does run something but requires to add more test coverage for those use cases. 

At the moment, the automation is in place, we need to add more test coverage to be in a more healthy place.

Not sure if I answered your concerns, please let me know If I need to clarify it.

**Comment by watson** [comment]
What I mean is that I think it's redundant. If I understand the process of how a new Docker image is being build, the 3 steps (`npm install ...`, `git add ...`, and `git commit ...`) doesn't do any difference in how the Docker image turns out as long as the Docker image is rebuilt, it will be built with a fresh `npm install` and hence get the latest version allowed in the range specified in `package.json`.

I would always recommend doing a manual upgrade between each major version of the agent, as it most likely involves manual checking and upgrading of code due to the breaking changes.

But maybe I misunderstood how the Docker images are built?

**Comment by kuisathaverat** [comment]
On the Docker, file installs the version that it is configured in the package.json, this means you have to manually edit this file before making an Opbeans release to set a fixed version for the Node.jas agent, the automated process that @v1v is putting in place make this change and trigger a push to build a new Docker image.

**Comment by watson** [comment]
In this repository, we don't have a checked-in `package-lock.json` file, so as long as you're running `npm install` from a freshly cloned version of the repo, you will get the latest version of the agent within the range specified in `package.json`. So if we specify `^3.0.0` in `package.json`, we'll get the latest `3.x` version.

You can verify this by running `npm ls elastic-apm-node` after finishing the `npm install`. That will show the actual version that has been installed.

**Comment by watson** [comment]
For details of what these prefixes to the version (like `^`) means in `package.json`, see: https://docs.npmjs.com/files/package.json#dependencies

**Pull request merged**